dotnet/src/Connectors/Connectors.GoogleVertexAI/Core/Gemini/Clients/GeminiChatCompletionClient.cs

```diff
@@ -0,0 +1,612 @@
+    }
+
+    /// <inheritdoc/>
+    public async Task<IReadOnlyList<ChatMessageContent>> GenerateChatMessageAsync(
+        ChatHistory chatHistory,
+        PromptExecutionSettings? executionSettings = null,
+        Kernel? kernel = null,
+        CancellationToken cancellationToken = default)
+    {
+        var state = ValidateInputAndCreateChatCompletionState(chatHistory, kernel, executionSettings);
+
+        for (state.Iteration = 1; ; state.Iteration++)
+        {
+            var geminiResponse = await this.SendRequestAndReturnValidGeminiResponseAsync(
+                    this._chatGenerationEndpoint, state.GeminiRequest, cancellationToken)
+                .ConfigureAwait(false);
+
+            var chatResponses = this.ProcessChatResponse(geminiResponse);
+
+            // If we don't want to attempt to invoke any functions, just return the result.
+            // Or if we are auto-invoking but we somehow end up with other than 1 choice even though only 1 was requested, similarly bail.
+            if (!state.AutoInvoke || chatResponses.Count != 1)
+            {
+                return chatResponses;
+            }
+
+            state.LastMessage = chatResponses[0];
+            if (state.LastMessage.ToolCalls is null)
+            {
+                return chatResponses;
+            }
+
+            // ToolCallBehavior is not null because we are in auto-invoke mode but we check it again to be sure it wasn't changed in the meantime
+            Verify.NotNull(state.ExecutionSettings.ToolCallBehavior);
+
+            state.AddLastMessageToChatHistoryAndRequest();
+            await this.ProcessFunctionsAsync(state, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<StreamingChatMessageContent> StreamGenerateChatMessageAsync(
+        ChatHistory chatHistory,
+        PromptExecutionSettings? executionSettings = null,
+        Kernel? kernel = null,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        var state = ValidateInputAndCreateChatCompletionState(chatHistory, kernel, executionSettings);
+
+        for (state.Iteration = 1; ; state.Iteration++)
+        {
+            using var httpRequestMessage = this.CreateHttpRequest(state.GeminiRequest, this._chatStreamingEndpoint);
+            using var response = await this.SendRequestAndGetResponseImmediatelyAfterHeadersReadAsync(httpRequestMessage, cancellationToken)
+                .ConfigureAwait(false);
+            using var responseStream = await response.Content.ReadAsStreamAndTranslateExceptionAsync()
+                .ConfigureAwait(false);
+
+            foreach (var messageContent in this.GetStreamingChatMessageContentsOrPopulateStateForToolCalling(state, responseStream))
```

Comment:
GetStreamingChatMessageContentsOrPopulateStateForToolCalling this method is synchronous, if you follow entire way to GeminiStreamJsonParser you will know why I didn't make this async.

Reply:
I actually started with GeminiStreamJsonParser, saw that it was synchronously reading character by character, which made me nervous, and I worked my way back here to say that it is indeed then doing synchronous I/O as part of an asynchronous method. We don't want to do that.

Reply:
I am going to take care for making this async tommorow.

Reply:
Cool. We can also address it subsequently if we want to get the PR as it is merged. It's just important to address once this is "real".

Reply:
Also, we should probably simplify method `GetStreamingChatMessageContentsOrPopulateStateForToolCalling` :). The name of this method identifies that it is doing a lot of work, maybe it's possible to split the logic.

Reply:
@dmytrostruk this is quite hard to refactor this method without performance impact. But method is not so complex indeed. This is version after update to async https://github.com/microsoft/semantic-kernel/blob/90b38059b129d936452242aa9af55b76d3623286/dotnet/src/Connectors/Connectors.GoogleVertexAI/Core/Gemini/Clients/GeminiChatCompletionClient.cs#L233

Reply:
@stephentoub I modified parser to async version. I not used same pattern like azure there [LINK](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/openai/Azure.AI.OpenAI/src/Helpers/SseAsyncEnumerator.cs) because It requires more work, now I'd like to merge gemini and back to this later